saleor/graphql/order/mutations/fulfillments.py

```diff
@@ -140,48 +201,88 @@ def clean_input(cls, data):
                 }
             )
 
+        lines_for_warehouses = defaultdict(list)
+        for line, order_line in zip(lines, order_lines):
+            for stock in line["stocks"]:
+                lines_for_warehouses[stock["warehouse"]].append(
+                    {"order_line": order_line, "quantity": stock["quantity"]}
+                )
+
         data["order_lines"] = order_lines
-        data["quantities"] = quantities
+        data["quantities"] = quantities_for_lines
+        data["lines_for_warehouses"] = lines_for_warehouses
         return data
 
     @classmethod
-    def save(cls, user, fulfillment, order, cleaned_input):
-        fulfillment.save()
-        order_lines = cleaned_input.get("order_lines")
-        quantities = cleaned_input.get("quantities")
+    def create_fulfillment(cls, fulfillment, warehouse_pk, lines):
         fulfillment_lines = []
-        for order_line, quantity in zip(order_lines, quantities):
+        for line in lines:
+            quantity = line["quantity"]
+            order_line = line["order_line"]
             if quantity > 0:
-                fulfill_order_line(order_line, quantity)
+                try:
+                    fulfill_order_line(order_line, quantity, warehouse_pk)
+                except InsufficientStock as exc:
+                    raise ValidationError(
+                        {
+                            "stocks": ValidationError(
+                                f"Insufficient product stock: {exc.item}",
+                                code=OrderErrorCode.INSUFFICIENT_STOCK,
+                            )
+                        }
+                    )
                 if order_line.is_digital:
                     order_line.variant.digital_content.urls.create(line=order_line)
                 fulfillment_lines.append(
                     models.FulfillmentLine(
                         order_line=order_line,
                         fulfillment=fulfillment,
                         quantity=quantity,
+                        stock=order_line.variant.stocks.get(warehouse=warehouse_pk),
                     )
                 )
+        return fulfillment_lines
 
-        fulfillment.lines.bulk_create(fulfillment_lines)
+    @classmethod
+    @transaction.atomic()
+    def create_fulfillments(cls, requester, order, cleaned_input):
```

Comment:
This function, as well as `create_fulfillment`, are doing business logic - IMO they should be moved out of mutation, most probably to `saleor/order/utils.py`. In `perform_mutation` we should only call this utility function and wrap it with `try..except` -> if any `InsufficientStock` error is raised we should raise the `ValidationError` then.